src/apps/vpn/controller.cpp

```diff
@@ -128,6 +128,56 @@ void Controller::initialize() {
     logger.info() << "Canary Ping Succeeded";
   });
 
+  connect(SettingsHolder::instance(), &SettingsHolder::transactionBegan, this,
```

Comment:
This file is already too "polluted". What would if in the CTOR you create a new object `ControllerTransitionMonitor` (a better name, please ;) that does this logic ?